File: 1901-find-a-peak-element-ii/1901-find-a-peak-element-ii.java
class Solution {
    private int findLargest(int[][] mat, int col) {
        int maxIndex = 0;
        for (int i = 1; i < mat.length; i++) {
            if (mat[i][col] > mat[maxIndex][col]) {
                maxIndex = i;
            }
        }
        return maxIndex;
    }
    
    public int[] findPeakGrid(int[][] mat) {
        int low = 0, high = mat[0].length - 1;
        while (low <= high) {
            int mid = low + ((high - low) >> 1);
            int largestIdx = findLargest(mat, mid);
            int left = mid - 1 >= 0 ? mat[largestIdx][mid - 1] : -1;
            int right = mid + 1 < mat[0].length ? mat[largestIdx][mid + 1] : -1;
            if (mat[largestIdx][mid] > left && mat[largestIdx][mid] > right) {
                return new int[]{largestIdx, mid};
            } else if (left > mat[largestIdx][mid]) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return new int[]{-1, -1};
    }
}

File: 1901-find-a-peak-element-ii/1901-find-a-peak-element-ii.cpp
class Solution {
public:
    int find_max(int col, vector<vector<int>>& mat, int n) {
        int idx = 0;
        for (int i = 0; i < n; i++) {
            if (mat[i][col] > mat[idx][col]) {
                idx = i;
            }
        }
        return idx;
    }
    
    vector<int> findPeakGrid(vector<vector<int>>& mat) {
        int n = mat.size(), m = mat[0].size();
        
        int low = 0, high = m - 1;
        
        while (low <= high) {
            int j = low + ((high - low) >> 1);
            
            int i = find_max(j, mat, n);
            
            int left = j - 1 >= 0 ? mat[i][j - 1] : -1;
            int right = j + 1 < m ? mat[i][j + 1] : -1;
            
            if (mat[i][j] > left && mat[i][j] > right) {
                return {i, j};
            }
            
            if (mat[i][j] < left) {
                high = j - 1;
            } else {
                low = j + 1;
            }
        }
        
        return {-1, -1};
    }
};